File: mobile/lib/features/tasks/models/page_response.dart
class PageResponse<T> {
  const PageResponse({
    required this.content,
    required this.page,
    required this.size,
    required this.totalElements,
    required this.totalPages,
  });

  final List<T> content;
  final int page;
  final int size;
  final int totalElements;
  final int totalPages;

  factory PageResponse.fromJson(
    Map<String, dynamic> json,
    T Function(Map<String, dynamic>) fromJsonT,
  ) {
    final rawContent = json['content'] as List<dynamic>? ?? const [];
    return PageResponse(
      content: rawContent
          .map((item) => fromJsonT(item as Map<String, dynamic>))
          .toList(),
      page: json['page'] as int? ?? 0,
      size: json['size'] as int? ?? rawContent.length,
      totalElements: json['totalElements'] as int? ?? rawContent.length,
      totalPages: json['totalPages'] as int? ?? 1,
    );
  }
}

File: backend/src/main/java/io/github/navysama/taskmanager/shared/web/PageResponse.java
package io.github.navysama.taskmanager.shared.web;

import java.util.List;
import org.springframework.data.domain.Page;

/**
 * Stable JSON shape for paginated lists. Spring Data's {@code Page} is never serialized directly: its JSON layout
 * is an implementation detail that is not guaranteed to stay stable.
 */
public record PageResponse<T>(List<T> content, int page, int size, long totalElements, int totalPages) {

    public static <T> PageResponse<T> from(Page<T> page) {
        return new PageResponse<>(
                page.getContent(), page.getNumber(), page.getSize(), page.getTotalElements(), page.getTotalPages());
    }
}
